### PR TITLE
Rewrite the link re-writing logic

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

The previous logic parses URLs with a regex first and then using
the markdown library. It then uses urlopen().read() to validate
links.

We use now the markdown library only to extract the list of links,
and then urlparse to deconstruct, analyse, adapt and reconstract
the link. We do not attempt to fetch links anymore, which means
that external links are not guaranteed to be working.

Absolute URLs are not changed (they may be external)
Fragments are relative to the page and do not need changes
Path only links should point to a file synced to the website
but sometimes the file may be missing (if it's not in the sync
configuration), so we follow this approach:
- prefix with base_path and check for the file locally
- if not found, prefix with base_url instead

Note that urlparse treats URLs without scheme like path only
URLs, so 'github.com' will be rewritten to base_url/github.com

Signed-off-by: Andrea Frittoli <andrea.frittoli@gmail.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/website/blob/master/CONTRIBUTING.md)
for more details._

/kind documentation

/cc @popcor255 @AlanGreene

### CODE REVIEW

**Comment by popcor255** [comment]
Under more examples -> https://deploy-preview-185--tekton.netlify.app/docs/pipelines/workspaces. The links do not work. I don't think we can get away with sync without checking if it exists. I wonder if it is possible to dns lookup or something less heavy then url open.

**Comment by popcor255** [comment]
The sync did not work properly. There is a list of bugs that reported the script generating broken links, #160, #126, #158, #133, #134. This refactor also includes a patch that fixes these bugs. The patch redirects a user to github if there is a relative link to a file that does not exist. This is important because some links will reference a snippet of code, test, or example.

These changes return it back to it's original state.
The logic follows this:

```
valid absolute links that are valid https url's -> stay the same
relative links links that are found locally -> stay the same
other links -> append https://github.com as a prefix 
```

**Comment by popcor255** [comment]
There are actually other extra links in trigger's page that are broken as well.

File: sync/sync.py [hunk 1]
```diff
@@ -58,76 +55,77 @@
...
+        return link
+    path = os.path.normpath(parsed.path)
+    if os.path.isfile(os.path.join(base_path, path)):
+        filename, ext = os.path.splitext(path)
+        # md files links are in the format .../[md filename]/
+        if ext == '.md':
+            path = filename + '/'
+        return urlunparse(parsed._replace(path="/".join([rewrite_path, path])))
+    # when not found on disk, append to the base_url
+    return urljoin(rewrite_url, link)
```
Comment (popcor255) [inline sync/sync.py:118]: This transform links does not transform links correctly.

**Review by popcor255 — CHANGES_REQUESTED** [review]

**Comment by popcor255** [comment]
Tests on master are also failing. o_O

**Comment by afrittoli** [comment]
Yeah, thanks for holding this, I found some broken links too, so I'm looking into those.

The problem is not so much with the changes in `is_url` though as that was only downloading content for absolute URLs anyways and paths would not be downloadable by `urlopen`. I added some logic in the rewrite process, for instance to rewrite `.md` links to a trailing slash to solve links like https://tekton.dev/docs/pipelines/container-contract/tasks.md (from https://tekton.dev/docs/pipelines/container-contract/) and something may have gone wrong in there.

**Comment by afrittoli** [comment]
> Under more examples -> https://deploy-preview-185--tekton.netlify.app/docs/pipelines/workspaces. The links do not work. I don't think we can get away with sync without checking if it exists. I wonder if it is possible to dns lookup or something less heavy then url open.

https://deploy-preview-185--tekton.netlify.app/docs/pipelines/workspaces/ works fine now

**Comment by afrittoli** [comment]
I still need to fix the last unit tests, but since we do not run those in CI yet, we could merge this, or we could wait and merge it on Monday. Since this would fix some broken links in the live site it might be good to merge it as it is and continue work on the unit tests as a follow up.

**Comment by afrittoli** [comment]
I fixed the unit tests now.
I think we should go ahead with this!

File: sync/sync.py (hunk 1 above)
Comment (afrittoli) [reply sync/sync.py:118]: Uhm, why not? Do you have an example?

**Comment by popcor255** [comment]
> I still need to fix the last unit tests, but since we do not run those in CI yet, we could merge this, or we could wait and merge it on Monday. Since this would fix some broken links in the live site it might be good to merge it as it is and continue work on the unit tests as a follow up.
> 
> I browsed around the preview and it looks ok. I found a couple of invalid links but they were broken in the source and it's not possible to fix them automatically, because the file is valid, but the fragment is not. See resources and workspaces under https://github.com/tektoncd/pipeline/blob/master/docs/tasks.md#reserved-directories.
> 
> The link in the preview is rendered correctly: https://deploy-preview-185--tekton.netlify.app/docs/pipelines/tasks/#resources but it does not take to the right location because resources is not a fragment on task.

If the fragment is invalid it should to redirect you to the page just not to the correct element. If you look under workspaces, the link that is given is `https://github.com/tektoncd/pipeline/tree/examples/v1beta1/taskruns/workspace.yaml` this returns a 404. The correct link would be `https://github.com/tektoncd/pipeline/tree/master/examples/v1beta1/taskruns/workspace.yaml`

**Pull request merged**